### PR TITLE
Fix  wss event loop panics on normal server shutdown/error return paths

### LINKED ISSUE #132: [BUG] transport review findings: shutdown deadlock, nil-listener panic, websocket leak, and reconnect lifecycle issues
## Summary

I reviewed the current `master` branch of `apache/dubbo-getty` locally and found several source-grounded lifecycle / concurrency / stability issues in the transport layer that look worth fixing.

The findings below are intentionally limited to issues with concrete code paths and clear impact, rather than broad speculation.

---

## 1. Client shutdown can deadlock because `client.stop()` closes sessions while holding `c.Lock()`, and session close can synchronously re-enter client reconnect logic

**Files**:
- `client.go`
- `session.go`

### Relevant code path

`client.stop()`:

```go
c.Lock()
for s := range c.ssMap {
    s.RemoveAttribute(sessionClientKey)
    s.RemoveAttribute(ignoreReconnectKey)
    s.Close()
}
c.Unlock()
```

`session.stop()`:

```go
clt, cltFound := s.GetAttribute(sessionClientKey).(*client)
ignoreReconnect, flagFound := s.GetAttribute(ignoreReconnectKey).(bool)
if cltFound && flagFound && !ignoreReconnect {
    clt.reConnect()
}
```

`reConnect()` calls methods that need the same client lock again (for example through `sessionNum()` / `connect()`).

### Why this matters

This creates a shutdown-time lock re-entry path:
- client shutdown holds `c.Lock()`
- session close runs synchronously under that lock
- session close can call back into `clt.reConnect()`
- reconnect path needs the same client lock again

That is a real deadlock risk during client shutdown.

### Suggested fix

- never call `s.Close()` while holding the client map lock
- first snapshot sessions under lock, then release the lock, then close them
- more generally, avoid synchronous reconnect work inside the session teardown path

---

## 2. Session teardown performs reconnect synchronously inside `session.stop()`, which makes close/shutdown paths unexpectedly heavy and fragile

**File**: `session.go`

### Relevant code path

```go
if cltFound && flagFound && !ignoreReconnect {
    clt.reConnect()
}
```

`reConnect()` itself can:
- loop,
- dial,
- sleep/backoff,
- and inspect/update client session state.

### Why this matters

A session close path is expected to be teardown-only, but here it can synchronously perform reconnection logic.

That means:
- `Close()` is no longer a small/local operation,
- teardown can block inside reconnect backoff/dial logic,
- and shutdown ordering becomes harder to reason about.

This also amplifies the deadlock problem above.

### Suggested fix

- decouple reconnect scheduling from `session.stop()`
- move reconnect decisions to a dedicated client lifecycle controller / background worker
- make teardown publish state and return, rather than performing reconnection inline

---

## 3. Deferred error path in `handlePackage()` can panic because of an incorrect nil check

**File**: `session.go`

### Relevant code path

```go
if err != nil {
    ...
    if s != nil || s.listener != nil {
        s.listener.OnError(s, err)
    }
}
```

The guard uses `||` instead of `&&`.

### Why this matters

If `s != nil` but `s.listener == nil`, the condition still passes and `s.listener.OnError(...)` dereferences a nil listener.

That means a normal I/O / decode / shutdown error in the deferred cleanup path can escalate into a panic.

### Suggested fix

- replace the guard with `if s != nil && s.listener != nil { ... }`
- add a regression test for the error-cleanup path with a nil listener

---

## 4. WebSocket self-connect rejection leaks the upgraded connection

**File**: `server.go`

### Relevant code path

```go
conn, err := s.upgrader.Upgrade(w, r, nil)
...
if conn.RemoteAddr().String() == conn.LocalAddr().String() {
    log.Warnf(...)
    return
}
```

After the websocket upgrade succeeds, the self-connect branch returns directly without closing `conn`.

### Why this matters

At that point the connection has already been upgraded and allocated. Returning without closing it leaks the socket/resource for that request path.

### Suggested fix

- explicitly close `conn` before returning in the self-connect rejection branch
- add a test or at least a small harness covering this branch if possible

---

## 5. WSS event loop panics on normal server shutdown/error return paths

**File**: `server.go`

### Relevant code path

```go
err = server.Serve(tls.NewListener(s.streamListener, config))
if err != nil {
    log.Errorf(...)
    panic(err)
}
```

The non-TLS WS loop logs serve errors and returns, but the WSS loop panics on any non-nil `Serve(...)` result.

### Why this matters

For long-running servers, `Serve(...)` returning is not always an exceptional condition. During normal shutdown, listener closure often causes `Serve` to return.

Treating that path as unconditional panic makes graceful shutdown brittle and can convert expected shutdown into process crash.

### Suggested fix

- distinguish expected shutdown errors from unexpected serve failures
- handle normal listener/server close without panic
- keep panic only for genuinely impossible internal invariants, if any

---

## 6. UDP receive buffer sizing contains a dead branch / likely logic bug

**File**: `session.go`

### Relevant code path

```go
maxBufLen = int(s.maxMsgLen + maxReadBufLen)
if int(s.maxMsgLen<<1) < bufLen {
    maxBufLen = int(s.maxMsgLen << 1)
}
```

At this point `bufLen` has not yet been populated from `conn.recv(buf)` and is still zero.

### Why this matters

The condition is effectively dead on entry and does not do what it appears to intend.

This looks like a real logic bug in UDP buffer sizing, and it likely means the branch never contributes to runtime behavior.

### Suggested fix

- re-check the intended sizing rule
- if the intent was to cap by `2 * maxMsgLen`, compare against the right value instead of uninitialized `bufLen`
- add a focused test around UDP buffer sizing / message length guard behavior

---

## 7. Reconnect attempt accounting appears monotonic across successes, which may permanently disable reconnect after enough intermittent failures

**File**: `client.go`

### Relevant code path

```go
var reconnectAttempts int
...
for {
    ...
    if connPoolSize <= sessionNum || maxReconnectAttempts < reconnectAttempts {
        break
    }
    c.connect()
    reconnectAttempts++
    ...
}
```

`reconnectAttempts` increases each loop, but I did not see it reset after a successful reconnect.

### Why this matters

If intermittent disconnect/reconnect cycles accumulate over process lifetime, the client may eventually stop reconnecting entirely once it crosses `maxReconnectAttempts`, even if previous reconnects actually succeeded.

### Suggested fix

- make retry accounting track consecutive reconnect failures rather than total lifetime attempts in the loop
- reset the counter when pool health is restored / a reconnect succeeds
- add a regression test for repeated disconnect-success-disconnect cycles

---

## Notes

I did **not** try to turn every suspicious pattern into an issue here. The items above are the ones that looked most actionable and had the clearest code evidence.

### PR DESCRIPTION
**Pull request by 222twotwotwo**
**What this PR does**:

This PR handles expected HTTP server shutdown errors in the WS/WSS event loops.

For WSS servers, normal shutdown or listener close errors are now treated as expected close paths instead of causing a panic. Unexpected WSS `Serve` errors still keep the original behavior: they are logged and then panic.

It also adds a regression test to verify that closing a WSS server does not panic during normal shutdown.

**Which issue(s) this PR fixes**:

Fixes #132 

**Special notes for your reviewer**:

The WSS panic behavior is preserved for unexpected `Serve` failures. This PR only excludes expected shutdown errors such as `http.ErrServerClosed`, `net.ErrClosed`, and closed listener errors.

Validated with:

```bash
go test ./...
```

**Does this PR introduce a user-facing change?**:

```release-note
WSS server shutdown no longer panics when the underlying HTTP server or listener is closed normally.
```

### CODE REVIEW

File: server_test.go [hunk 1]
```diff
@@ -226,6 +226,30 @@ func TestWSServeWSRequestClosesSelfConnectConn(t *testing.T) {
 	}
 }
 
+func TestWSSServerCloseDoesNotPanic(t *testing.T) {
+	certFile := filepath.Join(t.TempDir(), "server.crt")
+	keyFile := filepath.Join(t.TempDir(), "server.key")
```
Comment (AlexStocks) [inline server_test.go:234]: [P1] 补一条 unexpected Serve error 仍会失败的回归测试

当前用例只证明正常 `Close()` 不再触发 panic，没有保护本 PR 同时承诺保留的 fatal-error 路径。我在独立副本中把 `isHTTPServeCloseError` 变异为无条件返回 `true` 后，这个测试重复 20 次和全仓 `go test ./...` 仍全部通过，也就是所有 WSS Serve 故障被静默降级的回归无法被发现。请用可注入的 fake listener，必要时放到子进程中，分别验证 `http.ErrServerClosed`/wrapped `net.ErrClosed` 正常退出，而普通 fatal Accept/Serve error 必须进入 error 日志并保持 panic。

File: server.go [hunk 2]
```diff
@@ -421,11 +422,29 @@ func (s *server) runWSEventLoop(newSession NewSessionCallback) {
...
 }
 
+func (s *server) logHTTPServeClose(err error) {
+	log.Infof("http.server.Serve(addr{%s}) closed: %v", s.addr, err)
+}
+
+func isHTTPServeCloseError(err error) bool {
+	return errors.Is(err, http.ErrServerClosed) ||
+		errors.Is(err, net.ErrClosed) ||
+		strings.Contains(err.Error(), "use of closed network connection")
```
Comment (AlexStocks) [inline server.go]: [P2] 不要仅按错误文本判定正常关闭

这里的 substring 会把不属于关闭的真实故障一起吞掉。我用真实 `http.Server.Serve` 和故障 listener 返回 `fatal listener failure after use of closed network connection marker`，当前 helper 会判为正常关闭，WSS loop 因而跳过 `logHTTPServeError` 和 panic。仓库最低 Go 版本已能通过 `errors.Is(err, net.ErrClosed)` 识别标准 listener 关闭；建议移除文本匹配，或者至少与明确的 shutdown 状态/sentinel 链联合约束，并补对应分类测试。

Comment (222twotwotwo) [reply server.go]: 收到，已去除该文本匹配

File: server_test.go (hunk 1 above)
Comment (222twotwotwo) [reply server_test.go:234]: 收到，已补充

**Review by AlexStocks — APPROVED** [review]

**Comment by AlexStocks** [comment]
@222twotwotwo good work. I have merged this PR. please continue your work if there are other problems.
